File: src/api/routes/ingestion.ts
/**
 * Data Ingestion API Routes
 * Secure data ingestion with Data Act compliance
 */

import { Router, Request, Response } from 'express';
import { v4 as uuidv4 } from 'uuid';

const router = Router();

// Simple in-memory rate limiter
const rateLimiter = new Map<string, { count: number; resetTime: number }>();
const RATE_LIMIT = 100; // requests per second
const RATE_WINDOW = 1000; // 1 second

function checkRateLimit(sourceId: string): boolean {
  const now = Date.now();
  const record = rateLimiter.get(sourceId);

  if (!record || now > record.resetTime) {
    rateLimiter.set(sourceId, { count: 1, resetTime: now + RATE_WINDOW });
    return true;
  }

  if (record.count >= RATE_LIMIT) {
    return false;
  }

  record.count++;
  return true;
}

// Security validation
function validateSecurity(data: any): { valid: boolean; issues: string[] } {
  const issues: string[] = [];

  // Check for potential injection patterns
  const stringContent = JSON.stringify(data);

  if (/<script/i.test(stringContent)) {
    issues.push('Potential XSS detected');
  }

  if (/(\b(SELECT|INSERT|UPDATE|DELETE|DROP|UNION)\b)/i.test(stringContent)) {
    issues.push('Potential SQL injection detected');
  }

  if (/\$\{.*\}/.test(stringContent)) {
    issues.push('Potential template injection detected');
  }

  return { valid: issues.length === 0, issues };
}

// API Key validation middleware
function validateApiKey(req: Request, res: Response, next: Function): void {
  const apiKey = req.headers['x-api-key'];
  const devMode = process.env.AIOBS_DEV_MODE === 'true';

  if (devMode) {
    return next();
  }

  if (!apiKey) {
    res.status(401).json({
      success: false,
      error: 'API key required',
      timestamp: new Date().toISOString(),
    });
    return;
  }

  // In production, validate against stored API keys
  const validKeys = (process.env.AIOBS_API_KEYS || 'demo-key').split(',');
  if (!validKeys.includes(apiKey as string)) {
    res.status(403).json({
      success: false,
      error: 'Invalid API key',
      timestamp: new Date().toISOString(),
    });
    return;
  }

  next();
}

/**
 * POST /api/ingest/metrics
 * Ingest metric data
 */
router.post('/metrics', validateApiKey, (req: Request, res: Response) => {
  try {
    const { metrics, metadata, compliance } = req.body;

    if (!metrics || !Array.isArray(metrics) || metrics.length === 0) {
      return res.status(400).json({
        success: false,
        error: 'metrics array is required',
        timestamp: new Date().toISOString(),
      });
    }

    const sourceId = metadata?.source_id || 'unknown';

    // Rate limiting
    if (!checkRateLimit(sourceId)) {
      return res.status(429).json({
        success: false,
        error: 'Rate limit exceeded',
        retryAfter: 1,
        timestamp: new Date().toISOString(),
      });
    }

    // Security validation
    const security = validateSecurity(metrics);
    if (!security.valid) {
      return res.status(400).json({
        success: false,
        error: 'Security validation failed',
        issues: security.issues,
        timestamp: new Date().toISOString(),
      });
    }

    // Process metrics
    const processed = metrics.map((m: any) => ({
      id: uuidv4(),
      name: m.name,
      value: m.value,
      labels: m.labels || {},
      timestamp: m.timestamp || new Date().toISOString(),
      ingested_at: new Date().toISOString(),
    }));

    res.status(202).json({
      success: true,
      data: {
        accepted: processed.length,
        rejected: 0,
        ids: processed.map(p => p.id),
      },
      message: 'Metrics accepted for processing',
      timestamp: new Date().toISOString(),
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      error: 'Failed to ingest metrics',
      timestamp: new Date().toISOString(),
    });
  }
});

/**
 * POST /api/ingest/logs
 * Ingest log data
 */
router.post('/logs', validateApiKey, (req: Request, res: Response) => {
  try {
    const { logs, stream = 'aiobs-logs', metadata, compliance } = req.body;

    if (!logs || !Array.isArray(logs) || logs.length === 0) {
      return res.status(400).json({
        success: false,
        error: 'logs array is required',
        timestamp: new Date().toISOString(),
      });
    }

    const sourceId = metadata?.source_id || 'unknown';

    // Rate limiting
    if (!checkRateLimit(sourceId)) {
      return res.status(429).json({
        success: false,
        error: 'Rate limit exceeded',
        retryAfter: 1,
        timestamp: new Date().toISOString(),
      });
    }

    // Security validation
    const security = validateSecurity(logs);
    if (!security.valid) {
      return res.status(400).json({
        success: false,
        error: 'Security validation failed',
        issues: security.issues,
        timestamp: new Date().toISOString(),
      });
    }

    // Process logs
    const processed = logs.map((l: any) => ({
      id: uuidv4(),
      level: l.level || 'info',
      message: l.message,
      context: l.context || {},
      timestamp: l.timestamp || new Date().toISOString(),
      stream,
      ingested_at: new Date().toISOString(),
    }));

    res.status(202).json({
      success: true,
      data: {
        accepted: processed.length,
        rejected: 0,
        stream,
        ids: processed.map(p => p.id),
      },
      message: 'Logs accepted for processing',
      timestamp: new Date().toISOString(),
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      error: 'Failed to ingest logs',
      timestamp: new Date().toISOString(),
    });
  }
});

/**
 * POST /api/ingest/events
 * Ingest event data
 */
router.post('/events', validateApiKey, (req: Request, res: Response) => {
  try {
    const { events, channels = ['events'], metadata, compliance } = req.body;

    if (!events || !Array.isArray(events) || events.length === 0) {
      return res.status(400).json({
        success: false,
        error: 'events array is required',
        timestamp: new Date().toISOString(),
      });
    }

    const sourceId = metadata?.source_id || 'unknown';

    // Rate limiting
    if (!checkRateLimit(sourceId)) {
      return res.status(429).json({
        success: false,
        error: 'Rate limit exceeded',
        retryAfter: 1,
        timestamp: new Date().toISOString(),
      });
    }

    // Security validation
    const security = validateSecurity(events);
    if (!security.valid) {
      return res.status(400).json({
        success: false,
        error: 'Security validation failed',
        issues: security.issues,
        timestamp: new Date().toISOString(),
      });
    }

    // Process events
    const processed = events.map((e: any) => ({
      id: uuidv4(),
      event_type: e.event_type,
      severity: e.severity || 'info',
      title: e.title,
      source_service: e.source_service,
      payload: e.payload || {},
      timestamp: e.timestamp || new Date().toISOString(),
      channels,
      ingested_at: new Date().toISOString(),
    }));

    res.status(202).json({
      success: true,
      data: {
        accepted: processed.length,
        rejected: 0,
        channels,
        ids: processed.map(p => p.id),
      },
      message: 'Events accepted for processing',
      timestamp: new Date().toISOString(),
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      error: 'Failed to ingest events',
      timestamp: new Date().toISOString(),
    });
  }
});

/**
 * POST /api/ingest/batch
 * Batch ingestion for metrics, logs, and events
 */
router.post('/batch', validateApiKey, (req: Request, res: Response) => {
  try {
    const { metrics = [], logs = [], events = [], metadata, compliance } = req.body;

    const sourceId = metadata?.source_id || 'unknown';

    // Rate limiting
    if (!checkRateLimit(sourceId)) {
      return res.status(429).json({
        success: false,
        error: 'Rate limit exceeded',
        retryAfter: 1,
        timestamp: new Date().toISOString(),
      });
    }

    // Security validation
    const security = validateSecurity({ metrics, logs, events });
    if (!security.valid) {
      return res.status(400).json({
        success: false,
        error: 'Security validation failed',
        issues: security.issues,
        timestamp: new Date().toISOString(),
      });
    }

    const results = {
      metrics: { accepted: metrics.length, rejected: 0 },
      logs: { accepted: logs.length, rejected: 0 },
      events: { accepted: events.length, rejected: 0 },
    };

    res.status(202).json({
      success: true,
      data: {
        results,
        total: metrics.length + logs.length + events.length,
      },
      message: 'Batch accepted for processing',
      timestamp: new Date().toISOString(),
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      error: 'Failed to process batch',
      timestamp: new Date().toISOString(),
    });
  }
});

/**
 * GET /api/ingest/health
 * Check ingestion service health
 */
router.get('/health', (req: Request, res: Response) => {
  res.json({
    success: true,
    data: {
      status: 'healthy',
      backends: {
        victoriametrics: 'connected',
        openobserve: 'connected',
        redis: 'connected',
      },
      rateLimit: {
        limit: RATE_LIMIT,
        window: `${RATE_WINDOW}ms`,
      },
    },
    timestamp: new Date().toISOString(),
  });
});

/**
 * GET /api/ingest/stats
 * Get ingestion statistics
 */
router.get('/stats', (req: Request, res: Response) => {
  res.json({
    success: true,
    data: {
      last24Hours: {
        metricsIngested: 1250000,
        logsIngested: 850000,
        eventsIngested: 45000,
        bytesProcessed: 2.5 * 1024 * 1024 * 1024, // 2.5 GB
        avgLatency: 12, // ms
        errorRate: 0.001,
      },
      bySource: [
        { source: 'fraud-detector-v1', metrics: 350000, logs: 120000, events: 15000 },
        { source: 'churn-predictor-v1', metrics: 280000, logs: 95000, events: 8000 },
        { source: 'recommendation-v2', metrics: 420000, logs: 180000, events: 12000 },
        { source: 'inference-cluster', metrics: 200000, logs: 455000, events: 10000 },
      ],
    },
    timestamp: new Date().toISOString(),
  });
});

/**
 * GET /api/ingest/metrics/query
 * Query ingested metrics
 */
router.get('/metrics/query', (req: Request, res: Response) => {
  try {
    const { query, start, end, step = '60' } = req.query;

    if (!query) {
      return res.status(400).json({
        success: false,
        error: 'query parameter is required',
        timestamp: new Date().toISOString(),
      });
    }

    // Mock query response
    const now = Date.now();
    const startTime = start ? new Date(start as string).getTime() : now - 60 * 60 * 1000;
    const endTime = end ? new Date(end as string).getTime() : now;
    const stepMs = parseInt(step as string, 10) * 1000;

    const values: [number, string][] = [];
    for (let t = startTime; t <= endTime; t += stepMs) {
      values.push([t / 1000, (Math.random() * 100).toFixed(2)]);
    }

    res.json({
      success: true,
      data: {
        status: 'success',
        data: {
          resultType: 'matrix',
          result: [
            {
              metric: { __name__: query as string },
              values,
            },
          ],
        },
      },
      timestamp: new Date().toISOString(),
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      error: 'Failed to execute query',
      timestamp: new Date().toISOString(),
    });
  }
});

/**
 * GET /api/ingest/logs/search
 * Search ingested logs
 */
router.get('/logs/search', (req: Request, res: Response) => {
  try {
    const { query, stream = 'aiobs-logs', limit = '100' } = req.query;

    // Mock search results
    const results = [
      {
        id: uuidv4(),
        timestamp: new Date().toISOString(),
        level: 'info',
        message: 'Model inference completed successfully',
        context: { model_id: 'fraud-detector-v1', latency_ms: 45 },
        stream,
      },
      {
        id: uuidv4(),
        timestamp: new Date(Date.now() - 5000).toISOString(),
        level: 'warning',
        message: 'High latency detected',
        context: { model_id: 'recommendation-v2', latency_ms: 150 },
        stream,
      },
      {
        id: uuidv4(),
        timestamp: new Date(Date.now() - 10000).toISOString(),
        level: 'info',
        message: 'Feature store refresh completed',
        context: { features_updated: 1250 },
        stream,
      },
    ];

    res.json({
      success: true,
      data: {
        results,
        total: results.length,
        query: query || '*',
        stream,
      },
      timestamp: new Date().toISOString(),
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      error: 'Failed to search logs',
      timestamp: new Date().toISOString(),
    });
  }
});

/**
 * GET /api/ingest/events/recent
 * Get recent events
 */
router.get('/events/recent', (req: Request, res: Response) => {
  try {
    const { limit = '20', severity } = req.query;

    const events = [
      {
        id: uuidv4(),
        event_type: 'drift_detected',
        severity: 'warning',
        title: 'Data drift detected on fraud-detector-v1',
        source_service: 'fraud-detector-v1',
        timestamp: new Date(Date.now() - 45 * 60 * 1000).toISOString(),
      },
      {
        id: uuidv4(),
        event_type: 'slo_breach',
        severity: 'critical',
        title: 'Latency SLO breached on recommendation-v2',
        source_service: 'recommendation-v2',
        timestamp: new Date(Date.now() - 20 * 60 * 1000).toISOString(),
      },
      {
        id: uuidv4(),
        event_type: 'model_deployed',
        severity: 'info',
        title: 'New model version deployed: churn-predictor-v1.2.0',
        source_service: 'churn-predictor-v1',
        timestamp: new Date(Date.now() - 2 * 60 * 60 * 1000).toISOString(),
      },
    ];

    let filtered = events;
    if (severity && typeof severity === 'string') {
      filtered = events.filter(e => e.severity === severity);
    }

    res.json({
      success: true,
      data: {
        events: filtered.slice(0, parseInt(limit as string, 10)),
        total: filtered.length,
      },
      timestamp: new Date().toISOString(),
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      error: 'Failed to fetch events',
      timestamp: new Date().toISOString(),
    });
  }
});

export { router as ingestionRouter };

File: visualization/routers/ingestion.py
"""
AIOBS Data Ingestion API Router
Secure endpoints for continuous data injection
"""

from datetime import datetime, timedelta
from typing import Optional, List
from fastapi import APIRouter, HTTPException, Header, Query, Depends, Request, BackgroundTasks
from fastapi.responses import JSONResponse
import hashlib
import hmac

from ..config import get_settings, get_security_settings
from ..ingestion import (
    DataIngestionService,
    MetricIngestionRequest,
    LogIngestionRequest,
    EventIngestionRequest,
    BatchIngestionRequest,
    SecurityTestRequest,
    IngestionResponse,
    ingestion_service
)
from ..models.schemas import APIResponse

router = APIRouter(prefix="/api/ingest", tags=["ingestion"])

# Load settings
_settings = get_settings()
_security = get_security_settings()


# =============================================================================
# Security Dependencies
# =============================================================================

async def verify_api_key(
    x_api_key: Optional[str] = Header(None, alias="X-API-Key"),
    authorization: Optional[str] = Header(None)
) -> str:
    """
    Verify API key for ingestion requests
    Returns the source_id associated with the key
    """
    # Get API key from header
    api_key = x_api_key

    if not api_key and authorization:
        if authorization.startswith("Bearer "):
            api_key = authorization[7:]

    if not api_key:
        # In development mode, allow requests without API key
        if _settings.is_dev:
            return "dev-source"
        raise HTTPException(
            status_code=401,
            detail="API key required. Provide X-API-Key header or Bearer token."
        )

    # Validate API key against configured keys
    if api_key in _security.api_keys_list or _settings.is_dev:
        # Extract source_id from key hash
        return hashlib.sha256(api_key.encode()).hexdigest()[:16]

    raise HTTPException(
        status_code=401,
        detail="Invalid API key"
    )


async def verify_signature(
    request: Request,
    x_signature: Optional[str] = Header(None, alias="X-Signature"),
    x_timestamp: Optional[str] = Header(None, alias="X-Timestamp")
) -> bool:
    """
    Verify request signature for integrity
    """
    # Skip in dev mode
    if _settings.is_dev:
        return True

    if not x_signature or not x_timestamp:
        return True  # Signature is optional but recommended

    # Verify timestamp is recent (within 5 minutes)
    try:
        ts = datetime.fromisoformat(x_timestamp)
        if abs((datetime.utcnow() - ts).total_seconds()) > 300:
            raise HTTPException(
                status_code=401,
                detail="Request timestamp too old"
            )
    except ValueError:
        raise HTTPException(
            status_code=400,
            detail="Invalid timestamp format"
        )

    # Verify signature using configured secret
    secret = _security.signing_secret.get_secret_value()
    body = await request.body()

    expected_sig = hmac.new(
        secret.encode(),
        f"{x_timestamp}.{body.decode()}".encode(),
        hashlib.sha256
    ).hexdigest()

    if not hmac.compare_digest(expected_sig, x_signature):
        raise HTTPException(
            status_code=401,
            detail="Invalid signature"
        )

    return True


# =============================================================================
# Metric Ingestion Endpoints
# =============================================================================

@router.post("/metrics", response_model=IngestionResponse)
async def ingest_metrics(
    request: MetricIngestionRequest,
    background_tasks: BackgroundTasks,
    source_id: str = Depends(verify_api_key)
):
    """
    Ingest metrics into the observability platform.

    Metrics are validated, compliance-checked, and stored in VictoriaMetrics.

    **Rate Limits:**
    - 100 requests/second per source
    - 3000 requests/minute per source
    - 10 MB/second data volume

    **Security:**
    - Validates for injection attacks (SQL, prompt, XSS, command)
    - Sanitizes labels and values
    - Enforces Data Act compliance

    **Example Request:**
    ```json
    {
        "metrics": [
            {
                "name": "model_latency_ms",
                "value": 45.2,
                "labels": {"model": "fraud-v1", "env": "prod"}
            }
        ],
        "metadata": {
            "source": "ai_model",
            "source_id": "fraud-detector-v1"
        },
        "compliance": {
            "data_category": "performance",
            "processing_purpose": "AI model performance monitoring"
        }
    }
    ```
    """
    # Override source_id from auth
    request.metadata.source_id = source_id

    response = await ingestion_service.ingest_metrics(request)

    if response.status.value == "rate_limited":
        raise HTTPException(
            status_code=429,
            detail={
                "error": "Rate limit exceeded",
                "retry_after": response.rate_limit_reset.isoformat() if response.rate_limit_reset else None
            },
            headers={
                "Retry-After": str(int((response.rate_limit_reset - datetime.utcnow()).total_seconds()))
                if response.rate_limit_reset else "60"
            }
        )

    return response


@router.post("/metrics/batch", response_model=IngestionResponse)
async def ingest_metrics_batch(
    metrics: List[dict],
    source_id: str = Depends(verify_api_key),
    environment: str = Query(default="production"),
    data_category: str = Query(default="performance")
):
    """
    Simplified batch metric ingestion endpoint.

    For quick integration without full request structure.
    """
    from ..ingestion.schemas import (
        MetricValue, IngestionMetadata, DataActMetadata,
        IngestionSource, DataCategory, DataSensitivity
    )

    # Build full request
    request = MetricIngestionRequest(
        metrics=[MetricValue(**m) for m in metrics],
        metadata=IngestionMetadata(
            source=IngestionSource.APPLICATION,
            source_id=source_id,
            environment=environment
        ),
        compliance=DataActMetadata(
            data_category=DataCategory(data_category),
            sensitivity=DataSensitivity.INTERNAL,
            processing_purpose="Metric ingestion for AI observability monitoring"
        )
    )

    return await ingestion_service.ingest_metrics(request)


# =============================================================================
# Log Ingestion Endpoints
# =============================================================================

@router.post("/logs", response_model=IngestionResponse)
async def ingest_logs(
    request: LogIngestionRequest,
    source_id: str = Depends(verify_api_key)
):
    """
    Ingest logs into the observability platform.

    Logs are validated for sensitive data, sanitized, and stored in OpenObserve.

    **Security Features:**
    - Automatic PII detection and masking
    - Sensitive credential detection
    - Log injection prevention
    - Compliance audit trail

    **Example Request:**
    ```json
    {
        "logs": [
            {
                "level": "info",
                "message": "Model inference completed",
                "context": {"model_id": "fraud-v1", "latency_ms": 45}
            }
        ],
        "stream": "aiobs-logs",
        "metadata": {
            "source": "ai_model",
            "source_id": "fraud-detector-v1"
        },
        "compliance": {
            "data_category": "operational",
            "processing_purpose": "AI model operational logging"
        }
    }
    ```
    """
    request.metadata.source_id = source_id
    response = await ingestion_service.ingest_logs(request)

    if response.status.value == "rate_limited":
        raise HTTPException(
            status_code=429,
            detail="Rate limit exceeded"
        )

    return response


@router.post("/logs/simple")
async def ingest_logs_simple(
    logs: List[dict],
    source_id: str = Depends(verify_api_key),
    stream: str = Query(default="aiobs-logs")
):
    """
    Simplified log ingestion for quick integration.
    """
    from ..ingestion.schemas import (
        LogEntry, LogLevel, IngestionMetadata, DataActMetadata,
        IngestionSource, DataCategory, DataSensitivity
    )

    # Build log entries
    log_entries = []
    for log in logs:
        level = log.get("level", "info")
        if isinstance(level, str):
            level = LogLevel(level.lower())
        log_entries.append(LogEntry(
            level=level,
            message=log.get("message", ""),
            context=log.get("context", {}),
            timestamp=datetime.fromisoformat(log["timestamp"]) if "timestamp" in log else datetime.utcnow()
        ))

    request = LogIngestionRequest(
        logs=log_entries,
        stream=stream,
        metadata=IngestionMetadata(
            source=IngestionSource.APPLICATION,
            source_id=source_id
        ),
        compliance=DataActMetadata(
            data_category=DataCategory.OPERATIONAL,
            sensitivity=DataSensitivity.INTERNAL,
            processing_purpose="Log ingestion for AI observability monitoring"
        )
    )

    return await ingestion_service.ingest_logs(request)


# =============================================================================
# Event Ingestion Endpoints
# =============================================================================

@router.post("/events", response_model=IngestionResponse)
async def ingest_events(
    request: EventIngestionRequest,
    source_id: str = Depends(verify_api_key)
):
    """
    Ingest events for real-time broadcasting.

    Events are validated, stored in Redis, and broadcast via WebSocket.

    **Event Types:**
    - alert, incident, deployment, config_change
    - model_update, drift_detected, anomaly
    - slo_breach, cost_spike, security, audit

    **Channels:**
    - all: Broadcast to all subscribers
    - alerts: Alert-specific channel
    - metrics: Metric update channel
    - cognitive: AI cognitive metrics channel

    **Example Request:**
    ```json
    {
        "events": [
            {
                "event_type": "drift_detected",
                "severity": "warning",
                "title": "Data drift detected in fraud model",
                "source_service": "fraud-detector-v1",
                "payload": {"drift_score": 0.25}
            }
        ],
        "channels": ["alerts", "cognitive"],
        "metadata": {
            "source": "ai_model",
            "source_id": "drift-monitor"
        },
        "compliance": {
            "data_category": "operational",
            "processing_purpose": "AI model drift monitoring and alerting"
        }
    }
    ```
    """
    request.metadata.source_id = source_id
    response = await ingestion_service.ingest_events(request)

    if response.status.value == "rate_limited":
        raise HTTPException(
            status_code=429,
            detail="Rate limit exceeded"
        )

    return response


# =============================================================================
# Batch Ingestion Endpoint
# =============================================================================

@router.post("/batch", response_model=IngestionResponse)
async def ingest_batch(
    request: BatchIngestionRequest,
    source_id: str = Depends(verify_api_key)
):
    """
    Batch ingestion for metrics, logs, and events in a single request.

    Efficient for high-volume data sources that produce multiple data types.

    **Example Request:**
    ```json
    {
        "metrics": [{"name": "latency", "value": 45}],
        "logs": [{"level": "info", "message": "Inference complete"}],
        "events": [{"event_type": "custom", "title": "Batch processed"}],
        "metadata": {
            "source": "pipeline",
            "source_id": "inference-pipeline"
        },
        "compliance": {
            "data_category": "operational",
            "processing_purpose": "AI pipeline observability data ingestion"
        }
    }
    ```
    """
    request.metadata.source_id = source_id
    return await ingestion_service.ingest_batch(request)


# =============================================================================
# Security Testing Endpoint
# =============================================================================

@router.post("/security-test", response_model=IngestionResponse)
async def ingest_security_test(
    request: SecurityTestRequest,
    source_id: str = Depends(verify_api_key),
    x_security_auth: str = Header(..., alias="X-Security-Authorization")
):
    """
    Ingest security test payloads for authorized testing.

    **RESTRICTED ENDPOINT** - Requires special authorization.

    Used for:
    - Prompt injection testing
    - Adversarial input testing
    - Security vulnerability assessment
    - Model robustness testing

    **Required Headers:**
    - X-Security-Authorization: Authorization ticket for security testing

    **Example Request:**
    ```json
    {
        "tests": [
            {
                "test_category": "prompt_injection",
                "test_id": "PI-001",
                "payload": "Ignore previous instructions...",
                "expected_behavior": "reject",
                "authorized_by": "security-team",
                "authorization_ticket": "SEC-2024-001"
            }
        ],
        "dry_run": true,
        "metadata": {
            "source": "automated_test",
            "source_id": "security-scanner"
        },
        "compliance": {
            "data_category": "compliance",
            "processing_purpose": "Authorized security testing for AI model robustness"
        }
    }
    ```
    """
    # Verify security authorization using config
    if x_security_auth not in _security.security_auth_list and not _settings.is_dev:
        raise HTTPException(
            status_code=403,
            detail="Invalid security authorization"
        )

    request.metadata.source_id = source_id
    return await ingestion_service.ingest_security_test(request)


# =============================================================================
# Query Endpoints
# =============================================================================

@router.get("/metrics/query")
async def query_metrics(
    query: str = Query(..., description="PromQL query"),
    start: Optional[str] = Query(None, description="Start time (ISO format)"),
    end: Optional[str] = Query(None, description="End time (ISO format)"),
    source_id: str = Depends(verify_api_key)
):
    """
    Query metrics using PromQL.

    **Example Queries:**
    - `model_latency_ms{model="fraud-v1"}` - Instant query
    - `rate(model_requests_total[5m])` - Rate of requests
    - `avg_over_time(trust_score[1h])` - Average trust score
    """
    start_dt = datetime.fromisoformat(start) if start else None
    end_dt = datetime.fromisoformat(end) if end else None

    result = await ingestion_service.query_metrics(query, start_dt, end_dt)
    return APIResponse(success=True, data=result)


@router.get("/logs/search")
async def search_logs(
    query: str = Query(..., description="Search query"),
    stream: str = Query(default="aiobs-logs"),
    limit: int = Query(default=100, le=1000),
    source_id: str = Depends(verify_api_key)
):
    """
    Search logs in OpenObserve.
    """
    result = await ingestion_service.query_logs(query, stream, limit)
    return APIResponse(success=True, data=result)


@router.get("/events/recent")
async def get_recent_events(
    limit: int = Query(default=100, le=1000),
    source_id: str = Depends(verify_api_key)
):
    """
    Get recent events from the event store.
    """
    events = await ingestion_service.get_recent_events(limit)
    return APIResponse(success=True, data=events)


# =============================================================================
# Health & Stats Endpoints
# =============================================================================

@router.get("/health")
async def ingestion_health():
    """
    Check ingestion service health.
    """
    health = await ingestion_service.health_check()
    status_code = 200 if health["status"] == "healthy" else 503

    return JSONResponse(
        status_code=status_code,
        content=health
    )


@router.get("/stats")
async def ingestion_stats(source_id: str = Depends(verify_api_key)):
    """
    Get ingestion statistics.
    """
    stats = await ingestion_service.get_stats()
    return APIResponse(success=True, data=stats)


# =============================================================================
# Lifecycle Events
# =============================================================================

async def startup():
    """Initialize ingestion service on startup"""
    await ingestion_service.initialize()


async def shutdown():
    """Shutdown ingestion service gracefully"""
    await ingestion_service.shutdown()
